### PR TITLE
(feat) obsGroup count validation

### PR DESCRIPTION
**Pull request by arodidev**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary

This PR validates that the number of repeated obsGroup fields matches the repeat limit passed into the form's JSON 
## Screenshots

Screenshot 2023-12-01 at 11 01 36

### CODE REVIEW

**Review by pirupius — CHANGES_REQUESTED** [review]

File: src/components/encounter/ohri-encounter-form.component.tsx [hunk 1]
```diff
@@ -494,6 +497,20 @@ export const OHRIEncounterForm: React.FC<OHRIEncounterFormProps> = ({
       };
     }
 
+    if (encounterForSubmission.form.uuid === '1e5614d6-5306-11e6-beb8-9e71128cae77') {
```
Comment (kajambiya) [inline src/components/encounter/ohri-encounter-form.component.tsx:500]: Form Engine should not know about the `form uuid` It is possible for another implementation to have a different Uuid for the same form

File: src/utils/ohri-form-helper.ts [hunk 2]
```diff
@@ -1,27 +1,57 @@
...
+export const validateLNDbirthCount = (encounter: OpenmrsEncounter) => {
+  if (!encounter) {
+    const errorMessage = 'Missing encounter';
+    console.error(errorMessage);
+    return errorMessage;
+  }
+
+  const { obs: obsArray } = encounter;
+  const birthInfoObs = obsArray.filter((eachObs) => eachObs.concept === '1c70c490-cafa-4c95-9fdd-a30b62bb78b8');
+  const [birthCount] = obsArray.filter((eachObs) => eachObs.formFieldPath === 'ohri-forms-birth_count');
```
Comment (kajambiya) [inline src/utils/ohri-form-helper.ts:20]: FE shouldn't know about concepts and form field paths as these are bound to be implementation specific

File: src/utils/ohri-form-helper.ts [hunk 3]
```diff
@@ -1,27 +1,57 @@
...
 import { fetchConceptNameByUuid } from '../api/api';
 import { ConceptTrue } from '../constants';
 import { EncounterContext } from '../ohri-form-context';
-import { OHRIFormField, OHRIFormPage, OHRIFormSection, SubmissionHandler } from '../api/types';
+import { OHRIFormField, OHRIFormPage, OHRIFormSection, OpenmrsEncounter, SubmissionHandler } from '../api/types';
 import { OHRIDefaultFieldValueValidator } from '../validators/default-value-validator';
 import { isEmpty } from '../validators/ohri-form-validator';
 import { isTrue } from './boolean-utils';
 
+export const validateLNDbirthCount = (encounter: OpenmrsEncounter) => {
```
[the remark this answers is not in the document]
Comment (arodidev) [reply src/utils/ohri-form-helper.ts]: Noted.

File: src/utils/ohri-form-helper.ts (hunk 2 above)
Comment (arodidev) [reply src/utils/ohri-form-helper.ts]: Noted.

File: src/utils/ohri-form-helper.ts [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import dayjs from 'dayjs';
-import { LayoutType } from '@openmrs/esm-framework';
+import { LayoutType, showToast } from '@openmrs/esm-framework';
 import { fetchConceptNameByUuid } from '../api/api';
 import { ConceptTrue } from '../constants';
 import { EncounterContext } from '../ohri-form-context';
-import { OHRIFormField, OHRIFormPage, OHRIFormSection, SubmissionHandler } from '../api/types';
+import { OHRIFormField, OHRIFormPage, OHRIFormSection, OpenmrsEncounter, SubmissionHandler } from '../api/types';
```
Comment (kajambiya) [inline src/utils/ohri-form-helper.ts:6]: You added `showToast` and `OpenmrsEncounter` imports but looks like you don't use them.

**Review by kajambiya — COMMENTED** [review]
Hi Jamie, Nice work. Left some minor review comments

File: src/api/types.ts [hunk 5]
```diff
@@ -311,3 +311,9 @@ export interface ReferencedForm {
   formName: string;
   alias: string;
 }
+
+export type TobsGroupCounter = {
```
Comment (pirupius) [inline src/api/types.ts:315]: This naming doesn't look intuitive enough especially with the `T` can rename it to `RepeatGroupCounter` or `ObsGroupCounter` or something else

File: src/components/encounter/ohri-encounter-form.component.tsx [hunk 6]
```diff
@@ -430,6 +436,22 @@ export const OHRIEncounterForm: React.FC<OHRIEncounterFormProps> = ({
             uuid: field.uuid,
             voided: false,
           };
+
+          //validate obs group count against limit
+          const limit = field.questionOptions.repeatOptions?.limit;
+          const counter = obsGroupCounter?.filter((eachItem) => eachItem.fieldId === field.id)[0]?.obsGroupCount;
```
[the remark this answers is not in the document]
Comment (arodidev) [reply src/components/encounter/ohri-encounter-form.component.tsx:442]: Count always defaults to 1 as is implemented in the repeating component, and I pass this to the obsGroupCounter state, and with any cloned obsGroup added the count is updated and passed to the obsGroupCounter at the beginning of the array. Not sure if this answers your question.

File: src/utils/ohri-form-helper.ts (hunk 4 above)
Comment (arodidev) [reply src/utils/ohri-form-helper.ts]: This is resolved.

File: src/api/types.ts (hunk 5 above)
Comment (arodidev) [reply src/api/types.ts]: This is resolved.

**Review by samuelmale — CHANGES_REQUESTED** [review]

File: src/components/encounter/ohri-encounter-form.component.tsx [hunk 7]
```diff
@@ -430,6 +436,22 @@ export const OHRIEncounterForm: React.FC<OHRIEncounterFormProps> = ({
...
+
+          if (limit && counter && counter !== Number(limit)) {
+            setIsSubmitting(false);
+            showToast({
+              description: 'obsGroup count does not match limit specified',
+              title: 'Invalid entry',
+              kind: 'error',
+              critical: true,
+            });
```
[the remark this answers is not in the document]
Comment (arodidev) [reply src/components/encounter/ohri-encounter-form.component.tsx:451]: This is noted.

[the remark this answers is not in the document]
Comment (arodidev) [reply src/components/encounter/ohri-encounter-form.component.tsx:451]: @samuelmale Work on this has been slated to be fixed under technical debt together with a refactor for all other applicable areas concerning this, and as such will not be resolved in this PR. 
cc: @eudson

[the remark this answers is not in the document]
Comment (ebambo) [reply src/components/encounter/ohri-encounter-form.component.tsx:451]: Correct @arodidev ... we agree that the standards need to be followed but lets not block this specific PR.

**Review by ebambo — APPROVED** [review]

**Review by pirupius — APPROVED** [review]

**Pull request merged**